### PR TITLE
KetherNFT

### LINKED ISSUE #48: Launch KetherNFT: The NFT wrapper for Kether ads
Meta-issue for getting the KetherNFT contract production-ready.

- [x] [KetherNFT.sol](https://github.com/thousandetherhomepage/ketherhomepage/blob/master/contracts/KetherNFT.sol): Core contract is done
- [x] [KetherNFTRenderer.sol](https://github.com/thousandetherhomepage/ketherhomepage/blob/master/contracts/KetherNFTRender.sol): tokenURI on-chain rendering is functional but needs improvement.
- [x] `allAds` selector to make frontend rendering possibly with one query
- [x] Decide on `adminDisableRenderUpgrade`?
- [x] ~Rinkeby deployment~ (we have one but need to do it fresh) 
- [x] Finalize tokenURI JSON structure
  - [x] `forceNSFW` should override `nsfw` rather than be independent, right?
  - [x] NSFW state not showing up on OpenSea. Need to change type?
- [x] Confirm `owner()` function matches OpenSea ownership requirements
- [x] Confirm using the wrong `_owner` address in step 3 of minting (`wrap(...)`) will revert
- [x] Rinkeby deployment (again)
- [x] Write Etherscan instructions
  - [x] Does seed need to be kept secret?
- [x] Mainnet deployment
- [x] Minting frontend

Quick summary of how it works:
1. `Kether.precompute(uint _idx, address _owner) returns (bytes32 salt, address predictedAddress)` generates a precommitted address that will act as an atomic escrow for changing the ownership of the original ad to the NFT contract _and_ minting a corresponding NFT to your wallet. This happens client-side, so it does not cost any gas.
2. `KetherHomepage.setAdOwner(uint256 _idx_, address _newOwner)`  transfers the ownership of the ad on the original contract to the precommitment address. This is on-chain.
3. `KetherNFT.wrap(uint _idx, address _owner)` produces an ephemeral contract (`FlashEscrow`) at the precommitted address which atomically transfers the ad to the custody of the new KetherNFT contract and mints the NFT to the original owner. This is also a transaction on-chain.

To summarize the summary:  2 on-chain transactions to convert an ad to an NFT.

Anyone can "upgrade" their old ad to a new NFT ad, but all the old ads continue to be valid in non-NFT form. New ads purchased through the NFT contract are automatically reflected on both contracts-somewhat more gas-expensive, but there's no extra NFT-minting steps.

### PR DESCRIPTION
**Pull request by shazow**
This PR adds a KetherNFT contract for converting KetherHomepage ad units into ERC721-compatible NFTs that retain all the same control properties of ad units.

- [x] Write tests to verify that it works
- [x] Resolve all the TODOs and FIXMEs
- [x] Deploy on Rinkeby: Deployed at https://rinkeby.etherscan.io/address/0x6b69b05a5478a2390493765b496f70a52fe7b816

Closes #48

### CODE REVIEW

File: contracts/KetherNFT.sol [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+import "./IKetherHomepage.sol";
+
+
+// TODO: Name this something really cool
+contract Wrapper {
+  constructor(address target, bytes memory payload) {
+    (bool success,) = target.call(payload);
+    require(success, "Wrapper: target call failed");
+
+    selfdestruct(payable(target));
```
Comment (mveytsman) [inline contracts/KetherNFT.sol:16]: Is this who we want to be transferring to? See my idea for the bailout, I think this should be the owner not the target

Comment (shazow) [reply contracts/KetherNFT.sol:16]: I don't think it particularly matters, since this address shouldn't have a balance anyway but it's a required field for selfdestruct. This creates an easy way to give KetherHomepage a tip for using the NFT dapp. :P

File: test/kethernft.js [hunk 2]
```diff
@@ -0,0 +1,156 @@
...
+      const [addr,,,,,link,image,title] = await KH.ads(idx);
+      expect(addr).to.equal(KNFT.address);
+      expect(link).to.equal("foo2");
+      expect(image).to.equal("bar2");
+      expect(title).to.equal("baaz2");
+    }
+
+  });
+
+  it('wrap to non-owner', async function() {
```
Comment (shazow) [inline test/kethernft.js:80]: @mveytsman Check this test out when you get a chance, are we okay with these edge-cases? (I think so but wanna confirm)

File: contracts/KetherNFT.sol [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+
+  function unwrap(uint _idx, address _newOwner) external {
+    require(ownerOf(_idx) == msg.sender, "KetherNFT: unwrap for sender that is not owner");
+
+
+    instance.setAdOwner(_idx, _newOwner);
+    _burn(_idx);
+  }
+
+  // FIXME: If we let people set their own tokenURI, do we need another NSFW flag option here?
```
Comment (shazow) [inline contracts/KetherNFT.sol:89]: We could also wrap `tokenURI` getter to check if ad is marked as nsfw and return a placeholder uri instead?

Comment (shazow) [reply contracts/KetherNFT.sol:89]: Mainly I'd like to avoid nft wallets/marketplaces banning kethernft just because someone made an obscene tokenURI.

File: contracts/KetherNFT.sol [hunk 4]
```diff
@@ -0,0 +1,117 @@
...
+    require(_getAdOwner(_idx) == address(this), "KetherNFT: owner needs to be KetherNFT after wrap");
+    _mint(_owner, _idx);
+  }
+
+  function unwrap(uint _idx, address _newOwner) external {
+    require(ownerOf(_idx) == msg.sender, "KetherNFT: unwrap for sender that is not owner");
+
+
+    instance.setAdOwner(_idx, _newOwner);
+    _burn(_idx);
```
Comment (mveytsman) [inline contracts/KetherNFT.sol:86]: Make sure we have a test that wraps unwraps and wraps again.

Comment (mveytsman) [inline contracts/KetherNFT.sol:86]: Possibly check if the new owner is the owner before we burn.

File: contracts/KetherNFT.sol [hunk 5]
```diff
@@ -0,0 +1,117 @@
...
+  ///  - ipfs://Qme7ss3ARVgxv6rXqVPiikMJ8u2NLgmgszg13pYrDKEoiu
+  /// Images should be valid PNG.
+  /// Content-addressable storage links like IPFS are encouraged.
+  function publish(uint _idx, string calldata _link, string calldata _image, string calldata _title, bool _NSFW) external {
+    require(_isApprovedOrOwner(msg.sender, _idx), "KetherNFT: publish for sender that is not approved");
+
+    instance.publish(_idx, _link, _image, _title, _NSFW);
+  }
+
+}
```
Comment (mveytsman) [inline contracts/KetherNFT.sol:117]: do we need a fallback function here?

File: contracts/KetherNFT.sol [hunk 6]
```diff
@@ -0,0 +1,117 @@
...
+
+    // Wrapper completes the transfer escrow atomically and self-destructs.
+    new Wrapper{salt: salt}(address(instance), _wrapPayload(_idx));
+
+    require(_getAdOwner(_idx) == address(this), "KetherNFT: owner needs to be KetherNFT after wrap");
+    _mint(_owner, _idx);
+  }
+
+  function unwrap(uint _idx, address _newOwner) external {
+    require(ownerOf(_idx) == msg.sender, "KetherNFT: unwrap for sender that is not owner");
```
Comment (mveytsman) [inline contracts/KetherNFT.sol:82]: note this isnt checking approvals.

Comment (mveytsman) [reply contracts/KetherNFT.sol:82]: Do we want to use `_msgSender()` and follow ERC2771  for this?

File: contracts/KetherNFT.sol [hunk 7]
```diff
@@ -0,0 +1,117 @@
...
+  /// The link and image must be full web3-recognizeable URLs, such as:
+  ///  - bzz://a5c10851ef054c268a2438f10a21f6efe3dc3dcdcc2ea0e6a1a7a38bf8c91e23
+  ///  - bzz://mydomain.eth/ad.png
+  ///  - https://cdn.mydomain.com/ad.png
+  ///  - https://ipfs.io/ipfs/Qme7ss3ARVgxv6rXqVPiikMJ8u2NLgmgszg13pYrDKEoiu
+  ///  - ipfs://Qme7ss3ARVgxv6rXqVPiikMJ8u2NLgmgszg13pYrDKEoiu
+  /// Images should be valid PNG.
+  /// Content-addressable storage links like IPFS are encouraged.
+  function publish(uint _idx, string calldata _link, string calldata _image, string calldata _title, bool _NSFW) external {
+    require(_isApprovedOrOwner(msg.sender, _idx), "KetherNFT: publish for sender that is not approved");
```
Comment (mveytsman) [inline contracts/KetherNFT.sol:112]: `_msgSender()`

File: contracts/KetherNFT.sol [hunk 8]
```diff
@@ -0,0 +1,117 @@
...
+    _mint(_owner, _idx);
+  }
+
+  function unwrap(uint _idx, address _newOwner) external {
+    require(ownerOf(_idx) == msg.sender, "KetherNFT: unwrap for sender that is not owner");
+
+
+    instance.setAdOwner(_idx, _newOwner);
+    _burn(_idx);
+  }
```
Comment (mveytsman) [inline contracts/KetherNFT.sol:87]: We may want recoverUnminted that can only be called by us

```solidity
function recoverUnminted(uint _idx, address _to) {
  require(_msgSender() == admin)
  require(!_exists(_idx))
  require(_getAdOwner(_idx) == address(this))
  instance.setAdOwner(_idx, to)
}
```

Comment (mveytsman) [reply contracts/KetherNFT.sol:87]: I'm 50/50 on this...

File: contracts/KetherNFT.sol (hunk 5 above)
Comment (shazow) [reply contracts/KetherNFT.sol:143]: Added test confirming we don't

```javascript
  it("should not allow eth receive", async function() {
    const {account1} = accounts;

    expect(
      account1.sendTransaction({
        to: KNFT.address,
        value: ethers.utils.parseEther("42.0")
      })
    ).to.be.reverted;
  });
```

File: contracts/KetherNFT.sol (hunk 8 above)
Comment (shazow) [reply contracts/KetherNFT.sol:95]: Added it for now.

**Comment by mveytsman** [comment]
Here's my TODOlist from yesterday @shazow 

- [ ] unsuported network in the bottom menu
- [x] error about provider
- [x] clicking buy triggers a vuex error. Instead it should prompt you to choose a wallet.
- [ ] Migrate static footer links (FAQ, Press, )
- Metamask
	- [ ] with metamask: handle ropsten and others
- [ ] Verify "show NSFW"
- [ ] Monitor events
- [ ] Resize dragging things don't appear
Questions
- Can we do prerendering via nuxt instead of using ketherstatic.

**Comment by shazow** [comment]
Merged.
